manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,740 @@
+
+\subsection{ss:mm_domains}{Domains}
+
+Each domain is equipped with a \emph{frontier}, which is a map from non-atomic
+locations to timestamps. Intuitively, each domain's frontier records, for each
+non-atomic locations, the latest write known to the thread. More recent writes
+may have occurred, but are not guaranteed to be visible.
+
+For example,
+
+\begin{verbatim}
+d1: [a -> t1; b -> t3; c -> t6]
+d2: [a -> t1; b -> t4; c -> t7]
+\end{verbatim}
+
+\noindent represents two domains "d1" and "d2" and their frontiers.
+
+\subsection{ss:mm_na_access}{Non-atomic accesses}
+
+Let us now define the semantics of non-atomic reads and writes. Suppose domain
+"d1" performs the read of "b". For non-atomic reads, the domains may read an
+arbitrary element of the history for that location, as long as it is not older
+than the timestamp in the domains's frontier. In this case, since "d1" frontier
+at "b" is at "t3", the read may return the value "3", "4" or "5".
+
+Suppose domain "d2" writes the value "10" to "c" ("c := 10"). We pick a new
+timestamp "t9" for this write such that it is later than "d2"'s frontier at
+"c". Note a subtlety here: this new timestamp may not be later than everything
+else in the history, but merely later than any other write known to the writing
+domain. Hence, "t9" may be inserted in "c"'s history either (a) between "t7"
+and "t8" or (b) after "t8". Let us pick the former option for our discussion.
+Since the new write appears after all the writes known by the domain "d2" to
+the location "c", "d2"'s frontier at "c" is also updated. The new state of the
+abstract machine is:
+
+\begin{verbatim}
+(* Non-atomic locations *)
+a: [t1 -> 1; t2 -> 2]
+b: [t3 -> 3; t4 -> 4; t5 -> 5]
+c: [t6 -> 5; t7 -> 6; t9 -> 10; t8 -> 7] (* new write at t9 *)
+
+(* Domains *)
+d1: [a -> t1; b -> t3; c -> t6]
+d2: [a -> t1; b -> t4; c -> t9] (* frontier updated at c *)
+\end{verbatim}
+
+\subsection{ss:mm_at_access}{Atomic accesses}
+
+Atomic locations carry not only values but also synchronization information. We
+model atomic locations as a pair of the value held by that location and a
+frontier. The frontier models the synchronization information, which is merged
+with the frontiers of threads that operate on the location. In this way,
+non-atomic writes made by one thread can become known to another by
+communicating via an atomic location.
+
+For example,
+
+\begin{verbatim}
+(* Atomic locations *)
+A: 10, [a -> t1; b -> t5; c -> t7]
+B: 5,  [a -> t2; b -> t4; c -> t6]
+\end{verbatim}
+
+\noindent shows two atomic variables "A" and "B" with values "10" and "5",
+respectively, and frontiers of their own. We use upper-case variable names to
+indicate atomic locations.
+
+During atomic reads, the frontier of the location is merged into that of the
+domain performing the read. For example, suppose "d1" reads "B". The read
+returns "5", and "d1"'s frontier merged with "B"'s frontier, choosing the later
```

Comment:
This sounds like the merged frontier is returned as the result of the read. Perhaps it should say that d1's frontier is updated from B's?